src/sentry/eventstream/kafka/postprocessworker.py

```diff
@@ -0,0 +1,134 @@
+from sentry.eventstream.kafka.protocol import (
+    get_task_kwargs_for_message,
+    get_task_kwargs_for_message_from_headers,
+)
+from sentry.tasks.post_process import post_process_group
+from sentry.utils import metrics
+from sentry.utils.batching_kafka_consumer import AbstractBatchWorker
+from sentry.utils.cache import cache_key_for_event
+
+logger = logging.getLogger(__name__)
+
+Message = Any
+_DURATION_METRIC = "eventstream.duration"
+_CONCURRENCY_OPTION = "post-process-forwarder:concurrency"
+
+
+@contextmanager
+def _sampled_eventstream_timer(instance: str) -> Generator[None, None, None]:
+    record_metric = random.random() < 0.1
+    if record_metric is True:
+        with metrics.timer(_DURATION_METRIC, instance=instance):
+            yield
+    else:
+        yield
+
+
+def _get_task_kwargs(message: Message) -> Optional[Mapping[str, Any]]:
+    use_kafka_headers = options.get("post-process-forwarder:kafka-headers")
+
+    if use_kafka_headers:
+        try:
+            with _sampled_eventstream_timer(instance="get_task_kwargs_for_message_from_headers"):
+                return get_task_kwargs_for_message_from_headers(message.headers())
+        except Exception as error:
+            logger.error("Could not forward message: %s", error, exc_info=True)
+            with metrics.timer(_DURATION_METRIC, instance="get_task_kwargs_for_message"):
+                return get_task_kwargs_for_message(message.value())
+    else:
+        with metrics.timer(_DURATION_METRIC, instance="get_task_kwargs_for_message"):
+            return get_task_kwargs_for_message(message.value())
+
+
+def _record_metrics(partition: int, task_kwargs: Mapping[str, Any]) -> None:
+    event_type = "transactions" if task_kwargs["group_id"] is None else "errors"
+    metrics.incr(
+        "eventstream.messages",
+        tags={"partition": partition, "type": event_type},
+    )
+
+
+def dispatch_post_process_group_task(
+    event_id: str,
+    project_id: int,
+    group_id: Optional[int],
+    is_new: bool,
+    is_regression: bool,
+    is_new_group_environment: bool,
+    primary_hash: Optional[str],
+    skip_consume: bool = False,
+) -> None:
+    if skip_consume:
+        logger.info("post_process.skip.raw_event", extra={"event_id": event_id})
+    else:
+        cache_key = cache_key_for_event({"project": project_id, "event_id": event_id})
+
+        post_process_group.delay(
+            is_new=is_new,
+            is_regression=is_regression,
+            is_new_group_environment=is_new_group_environment,
+            primary_hash=primary_hash,
+            cache_key=cache_key,
+            group_id=group_id,
+        )
+
+
+def _get_task_kwargs_and_dispatch(message: Message):
+    task_kwargs = _get_task_kwargs(message)
+    if not task_kwargs:
+        return None
+
+    _record_metrics(message.partition(), task_kwargs)
+    dispatch_post_process_group_task(**task_kwargs)
+
+
+class PostProcessForwarderWorker(AbstractBatchWorker):
```

Comment:
The post process forwarder commits the staged offsets when a partition is revoked. How do you plan to support that ?

Reply:
Have put down some of my thoughts here https://github.com/getsentry/sentry/pull/28677#discussion_r713391530

I deliberately broke up the PR so that I can focus on smaller problems. I want to work on the use cases you mentioned in the integration PR.

There might need to be additional changes which might be required for both the PostProcessForwarderWorker and BatchingKafkaConsumer based on how the integration works

Reply:
Actually you should be fine. 
The batching consumer calls the flush method when partitions are revoked.